tests/executables/executables_ft.cpp

```diff
@@ -0,0 +1,247 @@
+#include <utcommon.h>
+#include <gtest/gtest.h>
+
+#ifdef _WIN32
+#define PATH_TO_BIN(binary) binary ".exe"
+#define MKDIR(path) "if not exist " + path + " mkdir " + path
+#define RMDIR(path) "rmdir /S /Q " + path
+#define POPEN _popen
+#define PCLOSE _pclose
+#define SETENV(env, value) _putenv_s(env, value)
+#else
+#define PATH_TO_BIN(binary) "./" binary
+#define MKDIR(path) "mkdir -p " + path
+#define RMDIR(path) "rm -rf " + path
+#define POPEN popen
+#define PCLOSE pclose
+#define SETENV(env, value) setenv(env, value, 1)
+#endif
+
+const static std::string TEST_TEMP_DIR = "executables_test_temp";
+
+// run the specified executable and record the output to output_path
+// if output_path is empty, just output to stdout
+static int run_and_record_executable(const std::string &shell_command_path, const std::string &output_path)
+{
+    std::string formatted_command = shell_command_path;
+    if (!output_path.empty())
+    {
+        formatted_command += " 2>&1 > " + output_path;
+    }
+    // In Linux, forking a process causes the under buffers to be forked, too. So, because popen uses fork under the
+    // hood, there may have been a risk of printing something in both processes. I'm not sure if this could happen in
+    // this situation, but better safe than sorry.
+    std::cout.flush();
+    FILE *process_stream = POPEN(formatted_command.c_str(), "r");
+    if (!process_stream)
+    {
+        return EXIT_FAILURE; // if popen fails, it returns null, which is an error
+    }
+    return PCLOSE(process_stream);
+}
+
+/*
+ * This function is used to verify test output. Specifically, it takes a stream (for the test output being checked) and
+ * a list of regular expressions. This function ensures that each regular expression provided matches a line in the
+ * output, in order. For example:
+ *
+ * input_stream:
+ * The device is plugged in on port 8.
+ * Today is Wednesday.
+ * The device is ready for use.
+ *
+ * Case 1
+ * regexes: { "The device is plugged in on port [0-9]+.", "The device is ready for use." }
+ *
+ * Pass; the first regex is tested against the first line and it matches. The second regex is tested against the second
+ * line and it does not match. So, the line is passed over, and the second regex is tested against the third line and
+ * does match.
+ *
+ * Case 2
+ * regexes: { "Today is Monday.", "The device is plugged .*" }
+ *
+ * Fail; the first regex is tested against the first line. It doesn't match, so the first regex is tested against the
+ * third line. It passes, so we move on to the third line and the second regex, which don't match. Note the importance
+ * of ordering.
+ */
+static void test_stream_against_regexes(std::istream &input_stream, const std::vector<std::string> &regexes)
+{
+    auto regex_iter = regexes.cbegin();
+    // ensure that all regexes are matched before the end of the file, in this order
+    while (input_stream && regex_iter != regexes.cend())
+    {
+        std::string results;
```

Comment:
How do we exit this loop if the input stream has less lines than the regexes vector?

Reply:
Just tested to make sure, but `input_stream` is part of the `while` condition and since `getline` runs on each iteration, eventually we will reach the EOF and so `input_stream` will evaluate to `false` and the loop will exit